app/questionnaire/navigator.py

```diff
@@ -46,21 +46,55 @@ def build_path(blocks, block_id, answers, path):
         for rule in blocks[block_id_index]['routing_rules']:
             goto_rule = rule['goto']
             if 'when' in goto_rule.keys():
-                answer_index = goto_rule['when']['id']
-                if answer_index in answers:
-                    answer = answers[answer_index]
-                    if evaluate_rule(goto_rule, answer):
-                        return build_path(blocks, goto_rule['id'], answers, path)
-                else:
-                    return path
+                if 'id' in goto_rule['when']:
+                    result = evaluate_answer_rule(goto_rule, answers)
+
+                if 'meta' in goto_rule['when']:
+                    result = evaluate_metadata_rule(goto_rule, metadata)
+
+                if result:
+                    return build_path(blocks, goto_rule['id'], metadata, answers, path)
+
             elif 'id' in goto_rule.keys():
-                return build_path(blocks, goto_rule['id'], answers, path)
+                return build_path(blocks, goto_rule['id'], metadata, answers, path)
     elif block_id_index != len(blocks) - 1:
         next_block_id = blocks[block_id_index + 1]['id']
-        return build_path(blocks, next_block_id, answers, path)
+        return build_path(blocks, next_block_id, metadata, answers, path)
     return path
 
 
+def evaluate_answer_rule(goto_rule, answers):
```

Comment:
Seems like we're getting to the point now that there are so many rules they should live in another module.